### ISSUE TITLE
Grouping nested nodes together

### ISSUE THREAD

**Issue by arunkumarpro1** [almende/vis#3293]
Hi,

Is there any way to group nested nodes together under a super node in vis.

I want to achieve the following behavior.

![design](https://user-images.githubusercontent.com/23283801/28596633-7b2b4a48-714e-11e7-976c-0d89d2f1a5ca.png)

I want to nest the nodes Node 11, Node 12, Node 13 within Node 1. So, when I click on Node 1  (**diagram (A)**), I would like to display the nested nodes and also extend the edges to the nested nodes (as shown in (**diagram (B)**).

Could anyone suggest me an idea to implement the above feature using vis?

Thanks in advance.

**Comment by wimrijnders** [comment]
Well, at least diagram B can be done. Hiding nodes in a super node is called *clustering* in `vis.Network`, see for instance [this example](http://visjs.org/examples/network/other/clustering.html). Clicking the super node to display the contained nodes is also possible.

However, to display them in a way such as diagram B is currently not possible. The official party line would no be 'feel free to implement it yourself', but there appears to be growing interest for this kind of containment; see e.g. the recently reopened #2319 and #3283 from 2 days ago. So it may be worthwhile to seriously consider this functionality.

...but try as I might, I can not think of a way to make this work right now. I've just spent some serious thinking time on this, and all I succeeded doing was stare blankly at the screen. If I have any insights, hackish or otherwise, I will share them here.

**Comment by wimrijnders** [comment]
Near duplicate of #2319.

**Comment by arunkumarpro1** [comment]
@wimrijnders Thanks for your reply and a very good explanation. It is really very helpful.
 Please let me know if you find any hack to do the above feature.

**Comment by arunkumarpro1** [comment]
@wimrijnders  But I really could not understand how the issue [https://github.com/almende/vis/issues/2319](2319) is related to my problem.

**Comment by wimrijnders** [comment]
Oops 😊 Typo. It's #2139. Sorry, fixing above links as well.

**Comment by arunkumarpro1** [comment]
:-) Yea.. I got it :+1:

**Comment by arunkumarpro1** [comment]
@wimrijnders Another question. How to make the label **bold** by default?

Sorry. I knew that this question is nowhere related to this issue.

**Comment by bradh** [comment]
@arunkumarpro1 Perhaps you try this in CSS, independent of vis.js?

**Comment by arunkumarpro1** [comment]
@bradh Is it not possible to do that in font object in vis options?

**Comment by wimrijnders** [comment]
Setting default bold font is actually not possible. The best way I can think of to do it is set option:

```
nodes: {
  font: {
    multi: true
  }
}
```

And then add `<b>...</b>` tags around all label texts.

@bradh using CSS is not possible here; it's because networks are drawn to `canvas`, meaning it's converted to a graphic format.

**Comment by arunkumarpro1** [comment]
@wimrijnders Thanks a lot. It works.

**Comment by wimrijnders** [comment]
Coming back to the original issue, I really don't know how to make this work. If you do find a solution, please share.

**Comment by arunkumarpro1** [comment]
Even I am not sure about it. I am still working on it. Will let you know once I am able to complete it.

**Comment by wimrijnders** [comment]
Great. I'll let you know here if I have any insights on this as well.

**Comment by AndrewMut** [comment]
@arunkumarpro1 That is how I am going to do same thing:
        1.Get data of all selected nodes
        2.Get all positions of selected
        3.Create 2 arrays for X and Y
        4.Sort X by biggest number and smallest
        5.Repeat for Y
        6.Add nessesary value to positions to prevent drawing frame through the nodes
        7.Use afterDrawing to draw custom lines to frame them
It's not directly what are u looking for, but at least something to start from.

**Comment by wimrijnders** [comment]
@AndrewMut you know, that might actually work! Is that in the light of a currently open issue here?

Thanks for applying your brainpower on this, I am unable to solve everything put here. I really appreciate it when others the the effort. This is something that can benefit many people.

**Comment by AndrewMut** [comment]
@wimrijnders Yup, it's in light of it, after You opened my eyes on afterDrawing vis become much more flexible for me.

**Comment by wimrijnders** [comment]
Glad to be of use!

**Comment by arunkumarpro1** [comment]
@AndrewMut Thanks for sharing your insights. I will work on them and share my progress here.

**Comment by AndrewMut** [comment]
@arunkumarpro1 Thanks, would be interesting to see!

**Comment by arunkumarpro1** [comment]
@AndrewMut Also if you have any working code, can you share it here for reference?

**Comment by AndrewMut** [comment]
@arunkumarpro1   I do not have, but if u would need a help, let me know plz, I'll try to write something to help you with that.

**Comment by arunkumarpro1** [comment]
@AndrewMut Thank you very much. I am working on the clustering part right now. I might need some help in drawing the boundaries around the nodes. I will let you know once I start working on that.

**Comment by arunkumarpro1** [comment]
@wimrijnders Is there any way to cluster a single node? In my requirement even if there is only one node that satisfies the cluster requirement, I have to embed it within a cluster.

**Comment by wimrijnders** [comment]
> Is there any way to cluster a single node? In my requirement even if there is only one node that satisfies the cluster requirement, I have to embed it within a cluster

Yes, there is. Set option `clusterNodeProperties.allowSingleNodeCluster:  true` and then proceed as normal.
**Update:** But I think you found that out for yourself, since the comment with the question has disappeared 😄.

**Comment by arunkumarpro1** [comment]
@wimrijnders Anyways, thanks a lot for the answer. 
I totally forgot the that you will be notified by my comment :P 

**Another question:**  I could not find the answer this time :) 
Is there any way to get hold of the canvas context, other than the render events ('beforeDrawing'/'afterDrawing')?

**Comment by wimrijnders** [comment]
Officially no, it is not exposed in the API. However, this is `javascript` and nothing whatsoever can stop you from accessing it by force.

If you don't mind doing unsupported things (but this is very unlikely to change), and you really really need direct access, it's at `network.canvas.frame.canvas`.

Now, please forget that I was the one who told you this 😶 .

**Comment by arunkumarpro1** [comment]
@wimrijnders haha. I will never say this to anyone :no_mouth:  

Also, I am skeptic about the working of 'afterDrawing' render event. Initially, I assumed that it will be called only once after the drawing is completed on the canvas. But after implementing it, I observed that the event is being triggered multiple times during stabilization? 

Is my understanding right?

**Comment by wimrijnders** [comment]
Ooh that's new for me. Never realized it worked that way. You may have uncovered a bug here.
But it will have to wait I'm afraid; vacation is breaking over me.

**Comment by arunkumarpro1** [comment]
@wimrijnders Sorry for bugging you. Enjoy your vacation :+1: 

We can discuss this once you are back from vacation.

**Comment by arunkumarpro1** [comment]
Hi @wimrijnders @AndrewMut I did get a chance to work on this feature this week. This is what I have achieved till now.

Initial layout (with clustered nodes):

![image](https://user-images.githubusercontent.com/23283801/29146477-2f58c072-7d16-11e7-8111-aaf7e2adb353.png)

Cluster opened on node selection:

![image](https://user-images.githubusercontent.com/23283801/29146778-ee3477d8-7d17-11e7-8bc8-87cdd4cd1ec2.png)

All Clusters opened on zoom:

![image](https://user-images.githubusercontent.com/23283801/29146796-0dd29ba6-7d18-11e7-9685-c2c7850422e6.png)

**Comment by arunkumarpro1** [comment]
@wimrijnders I am more concerned on the overlapping arrows. I tried a lot of physics configurations (hierarchicalRepulsion) , the layout remains the same every time as follows:

![image](https://user-images.githubusercontent.com/23283801/29146857-6420d464-7d18-11e7-94ca-a00c11ebf952.png)

This is my options object:
 
let options = {
    nodes: {
        shape: 'dot',
        font: {
            color: 'black',
            size: 9.4, // px
            face: 'arial',
            background: 'none',
            strokeWidth: 0, // px
            strokeColor: '#ffffff',
            align: 'center',
            multi: false,
            vadjust: 0
        },
        size: 20
    },
    layout: {
        randomSeed: undefined,
        improvedLayout: true,
        hierarchical: {
            enabled: true,
            levelSeparation: 100,
            nodeSpacing: 100,
            treeSpacing: 200,
            blockShifting: true,
            edgeMinimization: true,
            parentCentralization: true,
            direction: 'UD', // UD, DU, LR, RL
            sortMethod: 'directed' // hubsize, directed
        }
    },
    autoResize: false,
    edges: {
        smooth: false,
        color: '#000000',
        width: 0.5,
        arrows: {
            to: {
                enabled: true,
                scaleFactor: 0.5
            }
        }
    },
    physics: {
        enabled: true,
        hierarchicalRepulsion: {
            centralGravity: 0.0,
            springLength: 75,
            springConstant: 0.01,
            nodeDistance: 120,
            damping: 0.09
        },
        stabilization: {
            enabled: true,
            iterations: 1000 
        }
    }
};

**Comment by arunkumarpro1** [comment]
Hi @wimrijnders .. Sorry to bug you again. Do you have any idea about how I can avoid overlapping the edges?

I tried making a lot of changes to the hierarchialRepulsion physics config. I also tried to reduce the length of the edges by adjusting the length field within edges option / spring length with in hierarchialRepulsion physics config. Both of them did not have any effect on the edges.

**Comment by wimrijnders** [comment]
I'm afraid there's no short-term solution for this. I will need to dive into this code to figure out how to improve this. I've also got other bugs to deal with. Sorry.

**Comment by wimrijnders** [comment]
I'm still waiting for the brainwave to hit me, that will fix this....

**Comment by arunkumarpro1** [comment]
Hi @wimrijnders , right now I will be glad if you could give some ideas about how to avoid those overlapping edges. The graph looks so clumsy as follows

![screen shot 2017-08-31 at 2 49 13 pm](https://user-images.githubusercontent.com/23283801/29942529-d10abd3e-8e5b-11e7-8d5a-55cfb14cb8df.png)

Is there anything which I can do for placing the child nodes closer to their parent nodes? If you look into the above picture, most of the child nodes are placed very far away from their parent nodes which ultimately leads to the overlapping of edges.

**Comment by wimrijnders** [comment]
OK, so the thing to keep in mind, is that currently in the drawing methods, the order is retained for nodes as well as edges. Thus the positioning within the network can be influenced by crafty ordering of nodes/edges within the `DataSet`.

So for example what you *could* do in the given image, is to place `ALL DAEMONS` in front of `DROPPED PACKETS`. Consecutive use of this should clean up the layout considerably.

If I look at a previous network here above, the network looks quite clean to me. The layout should be handled fine no problem on setup. Strange. Would you mind sharing the node/edge definitions as well? A cut-down version showing the funky edges is perfectly fine.

Sorry about the lack of interaction, I'm back on a project and have much less time available.

**Comment by arunkumarpro1** [comment]
@wimrijnders By any chance are you planning to implement this feature in the library?

**Comment by nfigay** [comment]
Hi @arunkumarpro1 ,

I'm trying doing similar thing.

Some rough ideas:

Did you explore the way cola.js is doing similar thing, in particular looking at http://marvl.infotech.monash.edu/webcola/examples/SucroseBreakdown.html? So far I understood, cola.js performance are below vis.js, in particular with huge graphs. But may be something can be imagine from their way of doing for nesting nodes of a cluster in a container, preventing the nodes to be too far away.

Other idea: can it be envisage to consider a nested box  as something being able to have some attraction on the nodes of the cluster (considering the centre of the nested box), and being consider by the attraction algorithm. Unlike a traditional node, its shape will be draw dynamically  using you algorithm, but it will be able to attract it's content, being as a consequence more "contracted".  Can it be done creating "containing" edges which will not be displayed?

I'm interest  to know how you draw the boxes of the clusters: Dynamic SVG? Other?

**Comment by arunkumarpro1** [comment]
Hi @nfigay. Thanks for your comment. I will have a look at the features provided by cola.js.

I don't use SVG to draw boxes. I use canvas draw methods to draw lines surrounding the nodes of the same cluster.

**Comment by nfigay** [comment]
Hi @arunkumarpro1 ,
thanks for you feedback.
Let me know if you find what you need in cola.js ;)

**Comment by alanchenchen** [comment]
@arunkumarpro1 Hi ! I have met the same problem as yours, would you please tell me how you show multiple nodes in a group? emmm,i wanna show multiple nodes in a moveable group.

**Comment by arunkumarpro1** [comment]
Hi @alanchenchen,

The steps have been clearly explained by @AndrewMut  in the previous comments. I have just implemented the same idea.

Let me know if you need anything else.

**Comment by alanchenchen** [comment]
Thanks for your fast reply. @arunkumarpro1 

Should i have to use the canvas to draw lines to frame multiple nodes by myself? Well,the group nodes could be moved ? I wanna to move the group but not  move the nodes one by one.

**Comment by alanchenchen** [comment]
Thanks so much! I'm very surprised that you reponse to me at the same time while i'm asking for help .:stuck_out_tongue_closed_eyes: Please don't mind my poor English.:rofl: since of i'm a Chinese graduate..

**Comment by alanchenchen** [comment]
@arunkumarpro1 Hi! I'm sorry to bother you .I'm in need now that i've implemented the lines surrounding selected nodes. But i have no idea how to set the group move together if i drag the group...Could i have to set the option 'Cluster'?

**Comment by kengcher** [comment]
Hi I was wondering if visNetwork would be able to replicate the layout like http://marvl.infotech.monash.edu/webcola/examples/SucroseBreakdown.html without the collapsing nodes feature?

**Comment by AndrewMut** [comment]
Seems like its vis under the hood of that graph

ср, 25 июл. 2018 г., 4:20 KC <notifications@github.com>:

**Comment by Vincent-CIRCL** [comment]
It's probably not vis under the hood, as it is largely mentioned on the given page that this is the output of a research project : http://users.monash.edu/~mwybrow/dunnart/

**Comment by Vincent-CIRCL** [comment]
If you want a similiar feature as requested by OP, you may look there. I implemented a similar idea : 
https://github.com/Vincent-CIRCL/visjs_classificator/commit/73d7b73db0f5b667e3bfd3eff99cd357d468fad5

![image](https://user-images.githubusercontent.com/47740166/57228231-994a8c80-7013-11e9-85e8-dc885f6dba90.png)
